nurse_scheduling.py

```diff
@@ -0,0 +1,205 @@
+# Minimum duty days 'min_duty_days' - the number of work days that each
+# nurse wants
+# to be scheduled. At present, each will do the minimum on average.
+# The parameter gamma's value suggested by Ikeda, Nakamura, Humble
+gamma = 0.3
+preference = 1
+min_duty_days = int(n_days/n_nurses)
+
+
+# Find index into 1D list for (nurse_index, day_index)
+def get_index(nurse_index, day_index):
+    return nurse_index * n_days + day_index
+
+
+# Inverse of get_index - given an index in a 1D list, return the nurse_index
+# and day_index
+def get_nurse_and_day(index):
+    nurse_index, day_index = divmod(index, n_days)
+    return nurse_index, day_index
+
+
+# Hard nurse constraint: no nurse works two consecutive days
+# It does not have Lagrange parameter - instead, J matrix
+# symmetric, real-valued interaction matrix J, whereas all terms are
+# a or zero.
+# composite indices i(n, d) and j(n, d) as functions of n and d
+# In other words, the indices are not n and d. They are kind of
+# (n+d)**2, and we need to take that into account. This seems to be
+# a clean way to do it.
+# J_i(n,d)j(n,d+1) = a and 0 otherwise.
+J = defaultdict(int)
+for nurse in range(n_nurses):
+    for day in range(n_days - 1):
+        nurse_day_1 = get_index(nurse, day)
+        nurse_day_2 = get_index(nurse, day+1)
+        J[nurse_day_1, nurse_day_2] = a
+
+# Q matrix assign the cost term, the J matrix
+Q = deepcopy(J)
+
+# Hard shift constraint: at least one nurse working every day
+# The sum is over each day.
+# This constraint tries to make (effort * sum(q_i)) equal to workforce,
+# which is set to a constant in this implementation, so that one nurse
+# is working each day.
+# Overall hard shift constraint:
+# lagrange_parameter * sum_d ((sum_n(effort * q_i(n,d)) - workforce) ** 2)
+#
+# with constant effort and constant workforce:
+# = lagrange_parameter * sum_d ( effort * sum_n q_i(n,d) - workforce ) ** 2
+# = lagrange_parameter * sum_d [ effort * effort * (sum_n q_i(n,d) ** 2)
+#                              - 2 effort * workforce * sum_n q_i(n,d)
+#                              + workforce * workforce ]
+# The constant term is moved to the offset, below, right before we solve
+# the QUBO
+#
+# Expanding and merging the terms:
+# lagrange_parameter * (effort * effort - 2 effort * workforce) *
+# sum_d sum_n q_i(n,d)
+# + lagrange_parameter * effort * effort * sum_d sum_m sum_n q_i(n,d) q_j(m, d) #
+for nurse_day_1 in range(size):
+    _, date_index = get_nurse_and_day(nurse_day_1)
+    # Diagonal term, without the workforce * workforce
+    Q[nurse_day_1, nurse_day_1] += lagrange_parameter * ((effort * effort) - (2 * workforce * effort))
+    for nurse_day_2 in range(size):
+        _, day_index_2 = get_nurse_and_day(nurse_day_2)
+        # Include only the same day, across nurses
+        if (date_index == day_index_2 and nurse_day_2 != nurse_day_1):
+            Q[nurse_day_1, nurse_day_2] += 2 * lagrange_parameter * effort * effort
```

Comment:
This code could be simplified if you broke it up into two parts. (1) handle the diagonal terms and (2) the off-diagonal terms. This way, you don't have to deal with the two-component if-condition; that if-condition is a bit hard to read, in my opinion.

I believe using nurse and day indices directly and letting your helper handle the compound indices makes the code easier to read. However, please let me know if you feel differently.

```
# Diagonal terms of hard-shift constraint
for nurse in range(n_nurses):
    for day in range(n_days):
        ind = get_index(nurse, day)
        Q[ind, ind] += lagrange_parameter * effort**2 ..

# Off-diagonal terms in hard-shift constraint
# Note: include only the same day, across nurses
for day in range(n_days):
    for nurse1 in range(n_nurses):
        for nurse2 in range(n_nurses):

            # Same nurse; ignore diagonal index case
            if nurse1 == nurse2:
                continue

            ind1 = get_index(nurse1, day)
            ind2 = get_index(nurse2, day)
            Q[ind1, ind2] += 2 * lagrange_parameter * ..
```

Reply:
I did do this. It is much cleaner-looking. Please have a look.